### PR TITLE
CCMSG-1014: Do not leak fs in io.confluent.connect.hdfs.wal.WALFile.Writer#Writer().

### PR DESCRIPTION
**Pull request by avocader**
## Problem
HDFS client has a cache - `org.apache.hadoop.fs.FileSystem#CACHE`.
It's not automatically cleaned, it requires an explicit cleanup.
In the HDFS sink connector, some actions create new entries in the cache and some clean entries from it(added comments in the code, to make review easier, will remove before merging).

`io.confluent.connect.hdfs.wal.FSWAL#FSWAL` does actions, that create cache entries:
- `append()`
- `acquireLease()`
- `apply()`
- `truncate()`

Whenever we call `org.apache.hadoop.fs.FileSystem#newInstance(java.net.URI, org.apache.hadoop.conf.Configuration)`, it creates one entry in `org.apache.hadoop.fs.FileSystem#CACHE` so it has to be closed when not needed anymore.

There is a `catch` clause in `io.confluent.connect.hdfs.wal.WALFile.Writer#Writer`, that closes `fs` in case there is `RemoteException`.
But, in some cases, it can be another exception(e.g. `java.io.EOFException`, when HDFS is down), we are not catching it and re-throwing an exception.

## Solution
Catch `Exception` instead of `RemoteException` so that we never leak `fs`.

##### Does this solution apply anywhere else?
- [ ] yes
- [ ] no

##### Testing done:
- [x] Unit tests
- [x] Integration tests
- [ ] System tests
- [ ] Manual tests

### CODE REVIEW

File: src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java
Comment (kpatelatwork) [inline src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java:102]: do we have to call close in the RemoteExceptionBlock also specially when we are throwing ConnectException?

Comment (kpatelatwork) [reply src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java:102]: Also what about below block where you acquire the writer but we are throwing    

```
 if (sleepIntervalMs >= WALConstants.MAX_SLEEP_INTERVAL_MS) {
      throw new ConnectException("Cannot acquire lease after timeout, will retry.");
}
```

File: src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java
Comment (dosvath) [inline src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java:819]: `wal.append` in the `FSWAL.java` class calls two methods on the writer
```
      writer.append(key, value);
      writer.hsync();
```
Does this create two cache entries? above you mention a single one. If two entries are created and only one is cleaned that could also mean accumulation over time.

File: src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java
Comment (dosvath) [reply src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java:102]: +1 that `close()` may need to be called in the `RemoteExceptionBlock` before throwing a `ConnectException`.

Comment (dosvath) [reply src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java:102]: One additional thought about calling `close()` here (that may not be an issue) is that `close()` also closes the reader if it's non-null, but in this `acquireLease()` method we only call methods on the writer.

**Review by dosvath — APPROVED** [review]
Thanks @avocader I think we're close on resolving this.

**Review by dosvath — CHANGES_REQUESTED** [review]
Approved by accident, pending comments above.

File: src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java
Comment (avocader) [reply src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java:819]: @dosvath calling `writer.append(key, value);` and `writer.hsync();` doesn't create 2 cache entries(the size of `org.apache.hadoop.fs.FileSystem.Cache#map` doesn't change when I step over these lines in debugger).
Calling `wal.append(WAL.beginMarker, "");` creates one cache entry, this is a stack trace that I was able to get while running `HdfsSinkTaskTest`:
```
getInternal:3310, FileSystem$Cache (org.apache.hadoop.fs)
getUnique:3270, FileSystem$Cache (org.apache.hadoop.fs)
newInstance:528, FileSystem (org.apache.hadoop.fs)
<init>:146, WALFile$Writer (io.confluent.connect.hdfs.wal)
createWriter:78, WALFile (io.confluent.connect.hdfs.wal)
acquireLease:74, FSWAL (io.confluent.connect.hdfs.wal) --> we create a writer here
append:57, FSWAL (io.confluent.connect.hdfs.wal)
createWALs:365, HdfsSinkTaskTest (io.confluent.connect.hdfs)
testSinkTaskStartWithRecovery:233, HdfsSinkTaskTest (io.confluent.connect.hdfs)
...
```
So, essentially, creating one writer creates a one cache entry, and leaving unclosed writers(e.g. we catch and re-throw and exception in `io.confluent.connect.hdfs.wal.FSWAL#acquireLease`) blows up the cache.

File: src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java
Comment (avocader) [reply src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java:102]: Good catch @kpatelatwork! You are right, whenever we call
```
writer = WALFile.createWriter(conf, Writer.file(new Path(logFile)),
                                        Writer.appendIfExists(true));
```
it creates a cache entry, and we should always close it if applicable in case we are throwing an exception.
I moved `close()` to a `finally` block.

@dosvath right, `acquireLease()` only creates `writer`. Do you think it makes sense to split `close()` into `closeReader()` and `closeWriter()` and call `closeWriter()` from `acquireLease()`?

File: src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java
Comment (kpatelatwork) [inline src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java]: Hey @avocader  moving to finally will close the reader/writer even when the lease is successfully acquired.  I do not see that in the old code.  Are we sure this is the right fix?

Comment (avocader) [reply src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java]: I deleted `close()` call, so not relevant anymore.

File: src/test/java/io/confluent/connect/hdfs/wal/FSWALTest.java
Comment (kpatelatwork) [inline src/test/java/io/confluent/connect/hdfs/wal/FSWALTest.java:80]: Based on your new fix, Do we need a new test that is testing the Writer instead of acquireLease.  We should leave this test, I mean this is a good test.

File: src/test/java/io/confluent/connect/hdfs/wal/FSWALTest.java
Comment (kpatelatwork) [inline src/test/java/io/confluent/connect/hdfs/wal/FSWALTest.java:90]: One other question is that you did found out the real issue with Wal file and suggested customer to fix the permissions, don't we need that fix also?

File: src/test/java/io/confluent/connect/hdfs/wal/FSWALTest.java
Comment (avocader) [reply src/test/java/io/confluent/connect/hdfs/wal/FSWALTest.java:77]: Added a similar test - `io.confluent.connect.hdfs.wal.WALFileTest#testHdfsIsDown`.

File: src/test/java/io/confluent/connect/hdfs/wal/FSWALTest.java
Comment (avocader) [reply src/test/java/io/confluent/connect/hdfs/wal/FSWALTest.java:87]: It's a generic issue, that could be caused by multiple reasons - lack of HDFS privileges, HDFS is down, etc.
The PR fixes the generic issue of leaking `fs` instance in case there's an exception, different from `RemoteException`.
I also logged a separate ticket to improve permissions validation - https://confluentinc.atlassian.net/browse/CCMSG-1297

**Review by dosvath — APPROVED** [review]
LGTM, thanks @avocader, let's get +1 from @kpatelatwork as well.

File: src/test/java/io/confluent/connect/hdfs/wal/FSWALTest.java
Comment (kpatelatwork) [reply src/test/java/io/confluent/connect/hdfs/wal/FSWALTest.java:77]: Thanks it look good

**Review by kpatelatwork — APPROVED** [review]
@avocader This PR looks good. Thanks a lot for your persistence on the issue.

LGTM

**Pull request merged**
